analyzers/src/SonarAnalyzer.Common/Helpers/AnalysisContextExtensions.cs

```diff
@@ -56,27 +56,34 @@ public static SyntaxTree GetSyntaxTree<TLanguageKindEnum>(this CodeBlockStartAna
         public static SyntaxTree GetSyntaxTree(this SemanticModelAnalysisContext context) =>
             context.SemanticModel.SyntaxTree;
 
-        // ToDo: by default, do not verify if the project is Test or not when reporting. This is already being done at rule registration.
-        // Only check if it's a Test project for classes with multiple rules.
-        // https://github.com/SonarSource/sonar-dotnet/issues/4173
-        public static void ReportDiagnosticWhenActive(this SyntaxNodeAnalysisContext context, Diagnostic diagnostic) =>
-            ReportDiagnostic(new ReportingContext(context, diagnostic), SonarAnalysisContext.IsTestProjectNoCache(context.Compilation, context.Options));
+        /// <param name="verifyScopeContext">Provide value for this argument only if the class has more than one SupportedDiagnostics.</param>
+        public static void ReportDiagnosticWhenActive(this SyntaxNodeAnalysisContext context, Diagnostic diagnostic, SonarAnalysisContext verifyScopeContext = null) =>
+            ReportDiagnostic(new ReportingContext(context, diagnostic, verifyScopeContext));
 
+        // SyntaxTreeAnalysisContext doesn't have a Compilation => verifyScopeContext is never needed, because IsAnalysisScopeMatching returns always true.
         public static void ReportDiagnosticWhenActive(this SyntaxTreeAnalysisContext context, Diagnostic diagnostic) =>
-            ReportDiagnostic(new ReportingContext(context, diagnostic), SonarAnalysisContext.IsTestProjectNoCache(null, context.Options));
+            ReportDiagnostic(new ReportingContext(context, diagnostic, null));
 
         public static void ReportDiagnosticWhenActive(this CompilationAnalysisContext context, Diagnostic diagnostic) =>
-            ReportDiagnostic(new ReportingContext(context, diagnostic), SonarAnalysisContext.IsTestProject(context));
+            ReportDiagnostic(new ReportingContext(context, diagnostic, null) { IsTestProject = SonarAnalysisContext.IsTestProject(context) });
```

Comment:
I don't like this dual approach (assign IsTestProject property vs provide `verifyScopeContext`), but given current APIs we have, this is the most simple way to do it.

I'll definitely improve this in the future.

Reply:
Alternative way would be to make `SonarAnalysisContext.TryGetValueDelegate` public and change respective `IsTestProject` delegate to public, but that's not nice either.

Reply:
why not pass directly `bool? isTestProject` to the `ReportingContext` constructor?

the decision can be done inside the `AnalysisContextExtensions` class

Like this, all these decisions are taken here, and `ReportingContext` doesn't need to know about `SonarAnalysisContext`. It at least simplifies the interface, and the logic will be in only one file (currently, it's in two).

Reply:
I wanted to avoid repetition of the `IsTestProject` evaluation as it's having 5 different set of arguments, but but it's waaay cleaner like this.